pkg/store/proxy.go

```diff
@@ -344,18 +406,51 @@ func startStreamSeriesSet(
 
 	wg.Add(1)
 	go func() {
+		isFirstRecv := true
 		defer wg.Done()
 		defer close(s.recvCh)
-
 		for {
-			r, err := s.stream.Recv()
+			frameTimeoutCtx := context.Background()
+			var cancel context.CancelFunc
+			if s.responseTimeout != 0 {
+				frameTimeoutCtx, cancel = context.WithTimeout(frameTimeoutCtx, s.responseTimeout)
```

Comment:
Could we move out the construction of this context out of this function? It should probably make things easier to understand. WDYT, @bwplotka? This part is really becoming complex :(

Reply:
What are your thoughts on this, @IKSIN ?

Reply:
moved to function